File: internal/protocol/protocol.go
// Package protocol 定义 y-sync v1 客户端/服务端共享的协议类型。
// 元数据（JSON）与内容（原始字节流）在此分离：类型只描述元数据。
package protocol

// Op 常量：变更/操作类型
const (
	OpPut    = "put"
	OpMkdir  = "mkdir"
	OpMove   = "move"
	OpUnlink = "unlink"
)

// Node 类型常量
const (
	TypeFile = "file"
	TypeDir  = "dir"
)

// LoginReq POST /api/v1/auth/login
type LoginReq struct {
	User       string `json:"user"`
	Password   string `json:"password"`
	DeviceName string `json:"device_name"`
}

// LoginResp 登录成功返回每设备独立 token。
type LoginResp struct {
	Token      string `json:"token"`
	UserID     int64  `json:"user_id"`
	DeviceID   int64  `json:"device_id"`
	DeviceName string `json:"device_name"`
}

// NodeInfo 服务端文件树中的一个节点。parent_id==0 表示位于用户根目录下。
// Path 是用户空间内的完整相对路径（仅作展示/子树过滤；身份是 ID）。
type NodeInfo struct {
	ID          int64  `json:"id"`
	ParentID    int64  `json:"parent_id"`
	Name        string `json:"name"`
	Type        string `json:"type"`
	Path        string `json:"path"`
	Size        int64  `json:"size"`
	MTime       int64  `json:"mtime"` // unix 毫秒
	ContentHash string `json:"content_hash,omitempty"`
}

// Change 变更日志中的一条记录。客户端凭 cursor 增量拉取。
// 语义：该 node_id 当前位于 Path（unlink 表示已从 Path 移除）。
// 目录移动/重命名会为目录自身及其每个后代各产生一条 move 记录。
type Change struct {
	Cursor      int64  `json:"cursor"`
	DeviceID    int64  `json:"device_id"` // 产生该变更的设备（客户端跳过自己的变更）
	NodeID      int64  `json:"node_id"`
	Op          string `json:"op"`
	Path        string `json:"path"`
	ParentID    int64  `json:"parent_id"`
	Name        string `json:"name"`
	Type        string `json:"type"`
	Size        int64  `json:"size"`
	MTime       int64  `json:"mtime"`
	ContentHash string `json:"content_hash,omitempty"`
}

// ChangesResp GET /api/v1/sync/changes
type ChangesResp struct {
	Cursor  int64    `json:"cursor"` // 服务端当前 head cursor（即使返回条数少于 limit）
	Changes []Change `json:"changes"`
}

// HeadResp GET /api/v1/sync/head
type HeadResp struct {
	Cursor int64 `json:"cursor"`
}

// Op POST /api/v1/ops 批量元数据操作（服务端按序、原子应用）。
// mkdir: ParentID+Name；put: NodeID>0 表示覆盖已有节点内容，否则在 ParentID/Name 新建
// （ContentHash 必须已通过 /content 上传）；move: NodeID 移动到 ParentID+Name；
// unlink: NodeID（目录则递归）。
type Op struct {
	Op          string `json:"op"`
	NodeID      int64  `json:"node_id,omitempty"`
	ParentID    int64  `json:"parent_id,omitempty"`
	Name        string `json:"name"`
	Type        string `json:"type,omitempty"`
	ContentHash string `json:"content_hash,omitempty"`
	Size        int64  `json:"size,omitempty"`
	MTime       int64  `json:"mtime,omitempty"`
}

// OpResult 与 Ops 一一对应。
type OpResult struct {
	Ok     bool   `json:"ok"`
	Error  string `json:"error,omitempty"`
	NodeID int64  `json:"node_id,omitempty"`
	Cursor int64  `json:"cursor,omitempty"`
}

// OpsResp POST /api/v1/ops
type OpsResp struct {
	Results []OpResult `json:"results"`
}

// DedupResp PUT /api/v1/content
type DedupResp struct {
	Hash  string `json:"hash"`
	Dedup bool   `json:"dedup"`
}

// TrashItem 回收站条目（GET /api/v1/trash）。
type TrashItem struct {
	ID        int64  `json:"id"`
	OrigPath  string `json:"orig_path"`
	Name      string `json:"name"`
	Type      string `json:"type"`
	Hash      string `json:"content_hash,omitempty"`
	Size      int64  `json:"size"`
	MTime     int64  `json:"mtime"`
	DeletedAt int64  `json:"deleted_at"`
}

// VersionItem 文件历史版本条目。
type VersionItem struct {
	ID      int64  `json:"id"`
	NodeID  int64  `json:"node_id"`
	Path    string `json:"path"`
	Hash    string `json:"content_hash"`
	Size    int64  `json:"size"`
	MTime   int64  `json:"mtime"`
	Created int64  `json:"created"`
}

// ShareInfo 分享链接（FR-H1）。
type ShareInfo struct {
	Token     string `json:"token"`
	Path      string `json:"path"`
	NodeID    int64  `json:"node_id"`
	HasPwd    bool   `json:"has_password"`
	ExpiresAt int64  `json:"expires_at,omitempty"`
	Created   int64  `json:"created"`
}

// UploadSessionResp POST /api/v1/uploads（分块上传会话，FR-S11）。
type UploadSessionResp struct {
	ID       string  `json:"id"`
	Received []int64 `json:"received"` // 已收到的 chunk 序号（断点续传查询用）
}

File: rust/ysync-core/src/protocol.rs
//! 协议类型：与 Go 端 internal/protocol 完全一致的 JSON 形状（单一事实来源的两端镜像）。
use serde::{Deserialize, Serialize};

pub const OP_PUT: &str = "put";
pub const OP_MKDIR: &str = "mkdir";
pub const OP_MOVE: &str = "move";
pub const OP_UNLINK: &str = "unlink";
pub const TYPE_FILE: &str = "file";
pub const TYPE_DIR: &str = "dir";

#[derive(Debug, Clone, Serialize)]
pub struct LoginReq {
    #[serde(rename = "user")]
    pub user: String,
    #[serde(rename = "password")]
    pub password: String,
    #[serde(rename = "device_name")]
    pub device_name: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct LoginResp {
    #[serde(rename = "token", default)]
    pub token: String,
    #[serde(rename = "user_id", default)]
    pub user_id: i64,
    #[serde(rename = "device_id", default)]
    pub device_id: i64,
    #[serde(rename = "device_name", default)]
    pub device_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct NodeInfo {
    #[serde(rename = "id", default)]
    pub id: i64,
    #[serde(rename = "parent_id", default)]
    pub parent_id: i64,
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "type", default)]
    pub kind: String,
    #[serde(rename = "path", default)]
    pub path: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
    #[serde(rename = "content_hash", default, skip_serializing_if = "String::is_empty")]
    pub content_hash: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Change {
    #[serde(rename = "cursor", default)]
    pub cursor: i64,
    #[serde(rename = "device_id", default)]
    pub device_id: i64,
    #[serde(rename = "node_id", default)]
    pub node_id: i64,
    #[serde(rename = "op", default)]
    pub op: String,
    #[serde(rename = "path", default)]
    pub path: String,
    #[serde(rename = "parent_id", default)]
    pub parent_id: i64,
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "type", default)]
    pub kind: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
    #[serde(rename = "content_hash", default)]
    pub content_hash: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ChangesResp {
    #[serde(rename = "cursor", default)]
    pub cursor: i64,
    #[serde(rename = "changes", default)]
    pub changes: Vec<Change>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct HeadResp {
    #[serde(rename = "cursor", default)]
    pub cursor: i64,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct Op {
    #[serde(rename = "op")]
    pub op: String,
    #[serde(rename = "node_id", skip_serializing_if = "is_zero")]
    pub node_id: i64,
    #[serde(rename = "parent_id", skip_serializing_if = "is_zero")]
    pub parent_id: i64,
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "type", skip_serializing_if = "String::is_empty", default)]
    pub kind: String,
    #[serde(rename = "content_hash", skip_serializing_if = "String::is_empty", default)]
    pub content_hash: String,
    #[serde(rename = "size", skip_serializing_if = "is_zero")]
    pub size: i64,
    #[serde(rename = "mtime", skip_serializing_if = "is_zero")]
    pub mtime: i64,
}

fn is_zero(v: &i64) -> bool {
    *v == 0
}

#[derive(Debug, Clone, Deserialize)]
pub struct OpResult {
    #[serde(rename = "ok", default)]
    pub ok: bool,
    #[serde(rename = "error", default)]
    pub error: String,
    #[serde(rename = "node_id", default)]
    pub node_id: i64,
    #[serde(rename = "cursor", default)]
    pub cursor: i64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct OpsResp {
    #[serde(rename = "results", default)]
    pub results: Vec<OpResult>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DedupResp {
    #[serde(rename = "hash", default)]
    pub hash: String,
    #[serde(rename = "dedup", default)]
    pub dedup: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct TrashItem {
    #[serde(rename = "id", default)]
    pub id: i64,
    #[serde(rename = "orig_path", default)]
    pub orig_path: String,
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "type", default)]
    pub kind: String,
    #[serde(rename = "content_hash", default)]
    pub content_hash: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
    #[serde(rename = "deleted_at", default)]
    pub deleted_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct VersionItem {
    #[serde(rename = "id", default)]
    pub id: i64,
    #[serde(rename = "node_id", default)]
    pub node_id: i64,
    #[serde(rename = "path", default)]
    pub path: String,
    #[serde(rename = "content_hash", default)]
    pub content_hash: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
    #[serde(rename = "created", default)]
    pub created: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UploadSessionResp {
    #[serde(rename = "id", default)]
    pub id: String,
    #[serde(rename = "received", default)]
    pub received: Vec<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShareInfo {
    #[serde(rename = "token", default)]
    pub token: String,
    #[serde(rename = "path", default)]
    pub path: String,
    #[serde(rename = "node_id", default)]
    pub node_id: i64,
    #[serde(rename = "has_password", default)]
    pub has_password: bool,
    #[serde(rename = "expires_at", default)]
    pub expires_at: i64,
    #[serde(rename = "created", default)]
    pub created: i64,
}

/// daemon 控制端点的 JSON 类型（与 Go internal/client/daemon*.go 对应）。
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct FolderStatus {
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "local_path", default)]
    pub local_path: String,
    #[serde(rename = "cursor", default)]
    pub cursor: i64,
    #[serde(rename = "files", default)]
    pub files: i64,
    #[serde(rename = "last_sync", default)]
    pub last_sync: String,
    #[serde(rename = "last_error", default)]
    pub last_error: String,
    #[serde(rename = "conflicts_total", default)]
    pub conflicts_total: i64,
    #[serde(rename = "paused", default)]
    pub paused: bool,
    #[serde(rename = "last_stats", default)]
    pub last_stats: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct StatusResp {
    #[serde(rename = "folders", default)]
    pub folders: Option<Vec<FolderStatus>>,
}

impl StatusResp {
    pub fn folders_or_empty(&self) -> Vec<FolderStatus> {
        self.folders.clone().unwrap_or_default()
    }
}

/// null 容忍：Go 端历史实现可能输出 null 数组。
pub fn null_to_vec<'de, D, T>(de: D) -> Result<Vec<T>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: serde::Deserialize<'de>,
{
    let opt: Option<Vec<T>> = Option::deserialize(de)?;
    Ok(opt.unwrap_or_default())
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Conflict {
    #[serde(rename = "folder", default)]
    pub folder: String,
    #[serde(rename = "rel", default)]
    pub rel: String,
    #[serde(rename = "copy_rel", default)]
    pub copy_rel: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
}
